internal/repository/repository.go

```diff
@@ -252,12 +272,40 @@ func (r *Repository) LookupBlobSize(id restic.ID, tpe restic.BlobType) (uint, bo
 	return r.idx.LookupSize(restic.BlobHandle{ID: id, Type: tpe})
 }
 
+func (r *Repository) getZstdEncoder() *zstd.Encoder {
+	r.allocEnc.Do(func() {
+		enc, err := zstd.NewWriter(nil)
+		if err != nil {
+			panic(err)
+		}
+		r.enc = enc
+	})
+	return r.enc
+}
+
+func (r *Repository) getZstdDecoder() *zstd.Decoder {
+	r.allocDec.Do(func() {
+		dec, err := zstd.NewReader(nil)
+		if err != nil {
+			panic(err)
+		}
+		r.dec = dec
+	})
+	return r.dec
+}
```

Comment:
Cosmetic issue (feel free to ignore), but these don't need to be methods on Repository. We could put compression in a submodule internal/repository/zstd with just two functions for the API:

```
var (
	allocEnc sync.Once
	allocDec sync.Once
	enc      *zstd.Encoder
	dec      *zstd.Decoder
)

func Compress(p []byte) []byte {
	allocEnc.Do(func() {
		w, err := zstd.NewWriter(nil)
		if err != nil {
			panic(err)
		}
		enc = w
	})
	return enc.EncodeAll(p, nil)
}

func Decompress(p []byte) []byte { /* defined similarly */ }
```

Reply:
I haven't looked into yet, but I wonder: is that all worth it? Does the zstd encoder/decoder take so much time initializing or does it use a lot of memory?

Reply:
zstd supports parallel compression of a single stream, so NewWriter launches a pool of goroutines. (I don't think NewReader does, though.)

Reply:
The decoder also starts goroutines, see [repository: Fix leaked decoder in StreamPacks](https://github.com/restic/restic/pull/3666/commits/22b1fe791f16b3597e706365b8af1759194f103b). The encoder allocates server MBs of buffers. Without preallocation the compression performance was a complete disaster. With a preallocated encoder the compression performance is fine.

The code in the PR doesn't use the parallel compression feature, for that we'd have to use the `Write`-based API. But as the archiver already has its own goroutines, there's no benefit from introducing lots of synchronization points between goroutines.